### ISSUE TITLE
FP16 support

### ISSUE THREAD

**Issue by SeanNaren** [SeanNaren/deepspeech.pytorch#55]
I've opened a branch called fp_16 that has a benchmark script for fp_16.

I haven't got official confirmation from anyone at NVIDIA, but it seems like consumer cards (post Titan X maxwell) have been nerfed thus perform poorly using FP16 unless it is the Tesla P100 (to differentiate both markets, more info [here](https://github.com/pytorch/pytorch/issues/520#issuecomment-274161725)).

Measure the memory usage as well as the time taken for the benchmark script on various hardware at default settings. @ryanleary would it be possible to get benchmark times using this script? IIRC you have Titan X Maxwells?

**Comment by SeanNaren** [comment]
Some results using the benchmark scripts available on the `fp_16` branch.

```

**Comment by ryanleary** [comment]
```
Titan X (Pascal) benchmark_f16.py
F32 precision: Average training loop 0.59s Forward: 0.22s Backward: 0.37s
F16 precision: Average training loop 1.36s Forward: 0.58s Backward: 0.78s

Titan X (Pascal) benchmark_f16_forward.py
F32 precision: Forward time 0.07s
F16 precision: Forward time 0.14s
```

I have a machine with 2x Titan X (Pascal). Above are results. `nvidia-smi` didn't return any different numbers when the fp32 vs fp16 portions of the benchmark were running. How did you get those memory numbers exactly?

**Comment by SeanNaren** [comment]
Was hacky, I should probably flip the two benchmark's around and see if that works... I just commented out the benchmark doing the FP32 stuff, so `nvidia-smi` just showed the FP16 memory usage. Will try fix that in the script long term!

There is something strange going on with the DGX-1; we should've expected better performance on FP16 than FP32. The fact we didn't is a bit strange. On all other consumer cards, I expect poorer performance (as describe [here](http://www.anandtech.com/show/10325/the-nvidia-geforce-gtx-1080-and-1070-founders-edition-review/5))

**Comment by aaronzira** [comment]
If it helps, here are my results from benchmarks with 4 x GTX 1080Ti FE:

```
benchmark_fp16.py:
F32 precision: Average training loop 0.68s Forward: 0.27s Backward: 0.39s
F16 precision: Average training loop 0.70s Forward: 0.29s Backward: 0.39s

benchmark_fp16_single.py:
F32 precision: Forward time 0.07s
F16 precision: Forward time 0.12s
```

**Comment by SeanNaren** [comment]
@aaronzira thanks for that, I'm going to modify the script to do the correct procedure for FP_16 to see convergence (store weights as FP32, do update to them then transfer).

**Comment by EgorLakomkin** [comment]
2x GTX 1080:

benchmark_fp16.py:
F32 precision: Average training loop 0.63s Forward: 0.20s Backward: 0.42s
F16 precision: Average training loop 0.69s Forward: 0.25s Backward: 0.44s

benchmark_fp16_single.py:
Average times for Forward pass for LSTM: 
F32 precision: Forward time 0.07s 
F16 precision: Forward time 0.13s

**Comment by SeanNaren** [comment]
Thanks guys for the scores, I'll write up an explanation as to why F16 is slower soon using cuDNN (also yall have damn awesome setups!)

**Comment by SeanNaren** [comment]
So some more information as to why we see strange performance using FP16:

The main bottleneck in the networks are the RNNs. cuDNN supports only pseudo-fp16 for RNNs, meaning the multiples are done in FP32, and the weights are stored in FP16. This is regardless of architecture (thus even the same performance on a P100, since it does not support true-fp16). The regressions are strange and a result of the cublas library.

TLDR: FP16 will be the same as FP32 going forward, slower for backprop.

**Comment by SeanNaren** [comment]
As a result of the above, I don't think FP16 is currently viable. When changes happen and volta comes around I'll rethink about implementing this :)

**Comment by vaibhav0195** [comment]
Hi I tried using this library's latest apex code with fp16 but didnt get much of the performance boost i was expecting. I have opend an issue here 
https://github.com/SeanNaren/deepspeech.pytorch/issues/402#issue-425254385
can anyone help please ?
thanks :)
